File: 77. Combinations.py
# timeout

class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        self.res = []
        self.combination(1, [], n, k)
        return self.res
    
    def combination(self, start, path, n, k):
        if len(path) == k:
            self.res += [path[:]]
            return
        for i in xrange(start, n+1):
            self.combination(i+1, path + [i], n, k)

File: 77. Combinations.cpp
class Solution {
private:
    vector<vector<int>> res;
public:
    vector<vector<int>> combine(int n, int k) {
        vector <int> path;
        combination(1, path, n, k);
        return res;
    }
    
    void combination(int start, vector <int> path, int n, int k)
    {
        if (path.size() == k)
        {
            res.push_back(path);
            return;
        }
        for (int i = start; i <= n; i++)
        {
            path.push_back(i);
            combination(i+1, path, n, k);
            path.pop_back();
        }
    }
};
